### ISSUE TITLE
Add Cmd.fromAsync with named arguments 

### ISSUE THREAD

**Issue by Zaid-Ajaj** [elmish/elmish#166]
### Description

I am always bugged when using `Cmd.ofAsync` not accepting a simple `Async<'T>` and requires a function instead with it's argument supplied separately. This makes for some ugly code when the async expression is simple i.e. just `Async<'t>` in which case you must create a function accepting unit as input and use that in the `Cmd.ofAsync` giving unit as the second parameter or when it accepts multiple parameters, in which case you need supply the arguments except for the last and *then* supply the last parameter as the second argument for `Cmd.ofAsync` or the user has to go out of his/her way and use tupled arguments which is only friction that can be avoided

Ideally, one would only provide `Async<'t>` using something like the following:

```fs
let delayed (n: int) = 
  async { 
      do! Async.Sleep 1000
      return n
  }

let delayedCmd = 
  Cmd.fromAsync {
      Value = delayed 42
      Success = fun n -> ReceivedNumber n // or simply Success = ReceivedNumber
      Error = fun ex -> FailedToReceiveNumber ex.Message
  }
``` 
This version is much better for readability, no ambiguity when providing the arguments. 

### Implementation
```fs
module Cmd =
    type AsyncValParams<'a, 'b> =
        { Value : Async<'a>
          Success : 'a -> 'b
          Error : exn -> 'b }
    
    let fromAsync (asyncParams : AsyncValParams<'a, 'b>) : Elmish.Cmd<'b> =
        Elmish.Cmd.ofAsync (fun () -> asyncParams.Value) () asyncParams.Success asyncParams.Error
```

Another suggestion is to use `Cmd.ofAsyncMsg` implemented in Fabulous which is even more simple and straighforward but leaves error hanlding to the user:
```fs
module Cmd = 
    let ofAsyncMsg (value: Async<'t>) : Cmd<'t> = 
        let bind dispatch = 
            async {
               let! evaluatedMsg = value
               dispatch evaluatedMsg 
            } 
            |> Async.StartImmediate 
      
        [bind] 
```
To be used as follows:
```fs
let timeout n msg = 
    let delayedMsg =  
      async {
        do! Async.Sleep n
        return msg
      }

    Cmd.ofAsyncMsg delayedMsg 
```

**Comment by et1975** [comment]
Let's break this down into separate arguments:
>  you need supply the arguments except for the last and then supply the last parameter as the second argument for Cmd.ofAsync

Arguments situation is exactly the same for all functions constructing the `Cmd`: `ofFunc`, `ofPromise`, etc. The distinction of course is that the `Async` is cold and therefore doesn't *need* the arguments, but if it does then it's in the same boat as the rest of them and making it inconsistent is, arguably, not making it better.

At any rate this feels like a breaking change for the purpose of personal aesthetics and I'd like to avoid those. I'd be open to a separate `Cmd.Async` submodule where you can implement it any way you like, although, why not release your own library that shadows elmish `Cmd` module? Personally I don't use `Async` in the SPAs anymore and I don't use elmish on .NET, but maybe @cmeeren has a preference?

> Cmd.ofAsyncMsg implemented in Fabulous... leaves error hanlding to the user

Don mentioned that he doesn't particularly value `Cmd` in general or guiding the user towards error handling, so I take Fabulous decision making with a pound of salt. As implemented, it leaves the error handling to the dispatch loop... not seeing why guiding the user towards this would be a good idea.

**Comment by Zaid-Ajaj** [comment]
> Arguments situation is exactly the same for all functions constructing the Cmd: ofFunc, ofPromise, etc. The distinction of course is that the Async is cold and therefore doesn't need the arguments, but if it does then it's in the same boat as the rest of them and making it inconsistent is, arguably, not making it better.

> At any rate this feels like a breaking change for the purpose of personal aesthetics

It's not about my personal aesthetics, the idea it going towards better readability of Elmish code. As I mentioned earlier, it adds friction to code, for example when having just `Async<'t>` then users need to introduce a `unit -> Async<'t>` for no reason other than to satisfy `Cmd.ofAsync` , this code would otherwise be straightforward if you only accepted `Async<'t>`. 

In case you think this would introduce an inconsistency among the library functions, then I agree that it is a "breaking" change, but maybe also a good opportunity to better the library functions and make them *all* readable and consistent. Since a version 3.0 is not yet stable and people will expect things to break anyways going from 2.x to 3.x (otherwise why even bother to update to major version 3.x?) 

> Don mentioned that he doesn't particularly value Cmd in general or guiding the user towards error handling, so I take Fabulous decision making with a pound of salt

I am not talking about Don's preferences, this is just an example of simple and straightforward way of converting `Async` to `Cmd`. Many times you can assume that `Async<'t>` won't be triggering any errors (for example all functions in [Fable.SimpleHttp](https://github.com/Zaid-Ajaj/Fable.SimpleHttp)) so error handling won't be needed instead of dispatching a `NoOp` message on errors needlessly

**Comment by et1975** [comment]
Thanks for thoughtful response. So here's what I'm seeing:
> instead of dispatching a `NoOp` message on errors needlessly

There's a couple of `ofFunc` variations that doesn't require either `onSuccess` or `onError` arguments, I'm open to adding async version of those.

>not talking about Don's preferences

But you are providing the example of where his preferences lead and I'm saying I'm not satisfied with the direction.

 >why even bother to update to major version

Because it already has the changes I consider breaking and I'm sure you're aware there's more on the way, but fair enough we can add more breaking changes, the question is that of impact. If one has to change the library because of the my changes, that's fairly low impact, on the other hand if someone has to essentially rewrite their app because something so fundamental has changed it's a tougher proposition and I don't want our users to get on JS mill where everything is different every 6 months.

>  towards better readability of Elmish code

I hope you are not claiming that introducing a record that forces named arguments is *objectively* better after suggesting that forcing tuples on people is wrong. And again, I'm open to all those extensions but in a separate module, leaving the top `Cmd` consistent and handling the errors by default.

**Comment by cmeeren** [comment]
I too find the existing `ofAsync` signature somewhat unnatural, coming from a .NET background. In particular the (last) argument being separated out seems very weird, because as @Zaid-Ajaj says, it only really applies if you have a "hot monad" returned from a single-parameter function (including a single tupled argument), which I find is generally not the case.

More importantly though, I find the way it forces error handling in a particular way and location to be problematic. Most of my `async` CEs cannot throw, because I catch relevant exceptions in the CEs and return relevant results (e.g. Result with some error value based on the exception). As `ofAsync` currently exists, I either have to add a "NoOp" case to my Msg, or simply pass `raise` as the `onError` argument. I always go for the latter, and I can't imagine I am the only one taking this shortcut. I currently have no problem letting the dispatch loop handle truly exceptional errors, though I haven't considered whether this has any significant drawbacks.

As far as a concrete suggestion goes: Instead of separating out a single/last parameter, I have always thought that it would be better to just have promise/task/async-returning thunks, so the signature of `ofAsync` et al. would be (without error handling)

```f#
(unit -> Async<'a>) -> ('a -> 'msg) -> Cmd<'msg>
```

For `async` in particular one could argue that just `Async<'a>` is better as the first argument, though that breaks consistency. I'm not convinced signature consistency between hot and cold monads actually matter that much in this particular case.

In any case, I am not in favor of passing a record with named arguments. It's syntactically more verbose without adding (IMHO) significant clarity, and seems even less idiomatic and heavyweight than having the extra argument.

**Comment by Zaid-Ajaj** [comment]
> Because it already has the changes I consider breaking and I'm sure you're aware there's more on the way, but fair enough we can add more breaking changes, the question is that of impact. If one has to change the library because of the my changes, that's fairly low impact, on the other hand if someone has to essentially rewrite their app because something so fundamental has changed it's a tougher proposition and I don't want our users to get on JS mill where everything is different every 6 months.

The original proposition is adding `Cmd.fromAsync` which is not a breaking change: public API is backwards-compatible and has zero impact except for the need for some docs

> after suggesting that forcing tuples on people is wrong

Not sure where you read that because my suggestion is not to replace `ofAsync` but instead add `fromAsync` to the mix for places where `ofAsync` doesn't make sense and when people actually care about readable code

**Comment by et1975** [comment]
Sorry, I thought you were suggesting to apply this as a pattern across all functions for consistency - that would be a nasty change.

>when people actually care about readable code

Get off that high horse and lets see the PR ;)

But in general I'd like to see more libraries that expose `Cmd` as an API, instead of expanding elmish primitives.

**Comment by Zaid-Ajaj** [comment]
> Get off that high horse and lets see the PR ;)

I thought you would never ask ;) 

One last thing based on @cmeeren comments, do we make error handling optional? i.e. 
```fs
type AsyncValParams<'a, 'b> =
    { Value : Async<'a>
      Success : 'a -> 'b
      Error : Option<exn -> 'b> }
```

**Comment by MangelMaxime** [comment]
> But in general I'd like to see more libraries that expose Cmd as an API, instead of expanding elmish primitives.

@et1975 I agree with you on this point, but for primitives like `Async` it makes sense to supports them inside Elmish IMO.

Library-specific commands should indeed be implemented in the library like `Thoth.Elmish.Toast` or `Elmish.Toastr` do.

> One last thing based on @cmeeren comments, does it make sense to make error handling optional? i.e.

@Zaid-Ajaj 

IHMO we should take this opportunity to provide the same support as for the function.

So this means:
- `ofAsync` -> we don't change this one 
- `performAsync` -> ask only for `ofSuccess`
- `attemptSync` -> ask only for `ofError`
- `fromAsync` -> the one you proposed since the beginning

Like that all the scenarios are covered. And I guess it's possible to make `performAsync` and `attemptAsync` "named friendly" by default if we consider it better :).

**Comment by cmeeren** [comment]
@Zaid-Ajaj But why a record? Why not just normal function application of one or more arguments, like we do absolutely everywhere else when programming in F#? Doesn't that fit well with the proposed functions by @MangelMaxime?

**Comment by Zaid-Ajaj** [comment]
@cmeeren Because of readability. @MangelMaxime  if possible one function should be enough so that wouldn't have to think too much about it ;)

**Comment by MangelMaxime** [comment]
If you propose to provide a function taking a record then I have the same opinion as @et1975 😊

> I hope you are not claiming that introducing a record that forces named arguments is objectively better after suggesting that forcing tuples on people is wrong. And again, I'm open to all those extensions but in a separate module, leaving the top Cmd consistent and handling the errors by default.

**Comment by cmeeren** [comment]
I too am of the opinion that records are a weird and non-standard way to do this, and would much rather write and read normal function arguments, like I am used to everywhere else. This goes doubly now that we are talking about "overloads"/variants, where having to specify some record fields as `None` to get different behaviour (I would argue it's essentially different functions in this specific context) is unnecessarily verbose and runs completely counter to readability IMHO.

**Comment by cmeeren** [comment]
Also, instantiating records that you haven't defined yourself is not the easiest thing to do from memory, unless tooling has good support for inferring the necessity of a record and auto-creating a record instantiation syntax (VS doesn't).

**Comment by Zaid-Ajaj** [comment]
@cmeeren @MangelMaxime Approaching philosophical territories ... yeah let's not go there :smile:

**Comment by et1975** [comment]
When three maintainers express their concern with record API perhaps one should listen?

I'm finding that from library author perspective not only records hard on the tooling and the user, but they also make it hard to evolve things. I'm in the process of making the `Program` opaque and its implementation - internal, but even with that I wish I didn't model the type with record. 

I want to outline something similar to what @MangelMaxime suggested, but have 3 modules under `Cmd`  and have all the same "constructors" in them:
```fsharp
module Cmd =
  module OfFunc =
    let either: f a onSuccess onFailure : Cmd
    let attempt f a onSuccess : Cmd
    let perform f a onFailure : Cmd
    let result msg : Cmd
  module OfPromise =
    let either f a onSuccess onFailure : Cmd
    let attempt f a onSuccess : Cmd
    let perform f a onFailure : Cmd
    let result Promise<msg> : Cmd
  module OfAsync =
    let either f a onSuccess onFailure : Cmd
    let attempt f a onSuccess : Cmd
    let perform f a onFailure : Cmd
    let result Async<msg> : Cmd
```
And obsolete the existing constructors, inlining the new implementations for now.

**Comment by MangelMaxime** [comment]
Is the `Of` prefix really adding value/meaning?

**Comment by Zaid-Ajaj** [comment]
> When three maintainers express their concern with record API perhaps one should listen?

@et1975 

Someone said: "Get off that high horse and lets see the PR ;)"
 
@cmeeren No concrete complains, do I really have to go through this? OK *sight*

- Not easy to remember? *three words*, documented with examples on the function
- "Overloaded" variants because of `Option<'t>`? does any function accepting option is now considered "overloaded" and that is unacceptable by default? serious?
- Records are "non-standard"? are they though, since when?  
- is "unnecessarily verbose" and runs completely counter to readability IMHO? Three words <> verbose, the unnecessarily part is very arguable plus if you are unhappy then just use `ofAsync`

@MangelMaxime 

> If you propose to provide a function taking a record then I have the same opinion as @et1975

Basically just saying "I don't like records" while I clearly mentioned that they are not enforced and it is added to the mix. Totally optional, well documented, not a breaking change.

**Comment by cmeeren** [comment]
@Zaid-Ajaj: I still stand by my statements and agree with @et1975 and @MangelMaxime. I can try to elaborate on why I don't like records in this particular context, and why several of your bullet points misrepresent what I wrote. but unless strongly pressed I will drop it here because 1) I don't think it adds more noise than relevance and clarity to this discussion, 2) such things are at least partly personal taste, and 3) I don't have too much time on my hands these days. :) The fact is that three of us are of the opinion that the record syntax detracts more than it adds.

@et1975: I like where you're going with your suggestion; it solves the error handling stuff. But I still find it weird to have to pass the final argument separately. As I said, an alternative is to accept a thunk instead of `f` and `a`, and for `async` preferably just a raw `async` value. But I can certainly live with it, though. 

Also I'm not convinced the separation into modules adds meaningful clarity for users vs. just having `func` and `async` in the function names somehow (though I'm not saying it is a bad idea).

But is `Promise` defined in the core Elmish library or its dependencies? If so, should Elmish really define/depend on Promise?

**Comment by MangelMaxime** [comment]
@cmeeren `Promise` is defined in the native/browser API directly.

For Fable it's under `Fable.Import.JS` namespace included in `Fable.Core` which is included by all the Fable project and library.

When working with "synchronous" function in JavaScript we are using `Promise`. So it makes sense to include support for it in `Fable.Elmish`

**Comment by cmeeren** [comment]
Sure! Doesn't make sense for all Elmish renderers though, such as Elmish.WPF (as well as possible future non-Fable/non-JS renderers). But it's certainly not in the way, so I don't really have any problem with it, apart from dragging along a few unneeded dependencies. :)

**Comment by MangelMaxime** [comment]
This repo is releasing 2 packages:

- `Fable.Elmish` for Fable runtime
- `Elmish` for .Net runtime

And we include `promise` or `task` support depending on the targetted runtime. Same for the dependencies if you target .Net runtime we don't include `Fable.Core` as a dependency. You can check it on the [nuget feed](https://www.nuget.org/packages/Elmish/2.0.3)
